sig-api-machinery/design-docs/admission-control-webhooks.md

```diff
@@ -0,0 +1,934 @@
+
+InitializerConfiguration will not join `admissionregistration.k8s.io/v1beta1` at this time.
+
+Any webhooks that register with v1alpha1 may or may not be surprised when they start getting Versioned data.  But we don't make any promises for Alpha, and detecting which API was used to register is too complex.
+
+
+## Order of Admission 
+
+The recommended order on kubernetes.io for `--admission-control` will be: `MutatingAdmissionWebhook,NamespaceLifecycle,LimitRanger,ServiceAccount,PersistentVolumeLabel,DefaultStorageClass,DefaultTolerationSeconds,ValidatingAdmissionWebhook,ResourceQuota`
+
+ This order has the following properties:
+
+
+
+*   ResourceQuota comes last, so that if prior ones reject a request, it won't increment quota.
+*   All other Static ones are in the order recommended by [the docs](https://kubernetes.io/docs/admin/admission-controllers/#is-there-a-recommended-set-of-plug-ins-to-use).  (which variously do mutation and validation) Preserves the behavior when there are no webhooks.
+*   Ensures dynamic mutations happen before all validations.
+*   Ensures dynamic validations happen after all mutations.
+*   Users don't need to reason about the static ones, just the ones they add.
+
+Validation continues to happen after all the admission controllers (e.g. after mutating webhooks, static admission controllers, and non-mutating admission controllers.)
+
+**TODO**: we should move ResourceQuota after Validation, e.g. as described in #1137.  However, this is a longstanding bug and likely a larger change than can be done in 1.9, and should be revisited along with a larger quota redesign.  Therefore, it is not a requirement for Beta,
+
+
+## Parallel vs Serial 
+
+The main reason for parallel is reducing latency due to round trip and conversion.  We think this can often mitigated by consolidating multiple webhooks shared by the same project into one.
+
+Reasons not to allow parallel are complexity of reasoning about concurrent patches, and CRD not supporting PATCH.
+
+`ValidatingAdmissionWebhook `is already parallel,  and there are no responses to merge.  Therefore, it stays parallel.
+
+`MutatingAdmissionWebhook `will run in serial, to ensure conflicts are resolved deterministically.
+
+The order is the sort order of all the WebhookConfigs, by name, and by index within the Webhooks list.  
+
+We don't plan to make mutating webhooks parallel at this time, but we will revisit the question in the future and decide before going to GA.If a strong need arises for parallel mutating webhooks, we will carefully consider adding a dependency graph.the API will allow adding a partial order later as follows:  Add a "parallelStage" field to WebhookConfig object.  Any objects for which this field is unset all run in serial first.  Then ones which have the phase set are run.  All the ones with the same value run in parallel, and different ones in the order based on the numeric (or lexicographical) order by values).
```

Comment:
Please leave this at, "We don't plan to make mutating webhooks parallel at this time, but we will revisit the question in the future and decide before going to GA".  I don't want to to get caught up trying to review a future design for something against a set of future problems.

Reply:
Yeah the thing you quoted is a suggestion I left; everything else is what the suggestion was replacing. My doc -> md converter just pasted both there.